Index: tools/clang/lib/Basic/Targets.cpp
===================================================================
--- tools/clang/lib/Basic/Targets.cpp
+++ tools/clang/lib/Basic/Targets.cpp
@@ -7724,9 +7724,9 @@
 
 static const unsigned TCEOpenCLAddrSpaceMap[] = {
     0, // Default
-    3, // opencl_global
-    4, // opencl_local
-    5, // opencl_constant
+    1, // opencl_global
+    3, // opencl_local
+    2, // opencl_constant
     // FIXME: generic has to be added to the target
     0, // opencl_generic
     0, // cuda_device
