Index: clang/lib/Basic/Targets/TCE.h
===================================================================
--- clang/lib/Basic/Targets/TCE.h
+++ clang/lib/Basic/Targets/TCE.h
@@ -32,9 +32,9 @@ namespace targets {
 
 static const unsigned TCEOpenCLAddrSpaceMap[] = {
     0, // Default
-    3, // opencl_global
-    4, // opencl_local
-    5, // opencl_constant
+    1, // opencl_global
+    3, // opencl_local
+    2, // opencl_constant
     0, // opencl_private
     // FIXME: generic has to be added to the target
     0, // opencl_generic
