Index: lib/Transforms/Utils/SimplifyCFG.cpp
===================================================================
--- lib/Transforms/Utils/SimplifyCFG.cpp	(revision 168046)
+++ lib/Transforms/Utils/SimplifyCFG.cpp	(working copy)
@@ -3539,7 +3539,8 @@
   assert(SI->getNumCases() > 1 && "Degenerate switch?");
 
   // Only build lookup table when we have a target that supports it.
-  if (!TTI || !TTI->getScalarTargetTransformInfo()->shouldBuildLookupTables())
+  if (!TTI || TTI->getScalarTargetTransformInfo() == NULL ||
+      !TTI->getScalarTargetTransformInfo()->shouldBuildLookupTables())
     return false;
 
   // FIXME: If the switch is too sparse for a lookup table, perhaps we could
