Index: tools/clang/lib/Basic/Targets.cpp
===================================================================
--- tools/clang/lib/Basic/Targets.cpp	(revision 250891)
+++ tools/clang/lib/Basic/Targets.cpp	(working copy)
@@ -5974,8 +5974,10 @@
       FloatFormat = &llvm::APFloat::IEEEsingle;
       DoubleFormat = &llvm::APFloat::IEEEsingle;
       LongDoubleFormat = &llvm::APFloat::IEEEsingle;
-      DescriptionString = "E-p:32:32-i8:8:32-i16:16:32-i64:32"
-                          "-f64:32-v64:32-v128:32-a:0:32-n32";
+      DescriptionString = "E-p:32:32:32-i1:8:8-i8:8:32-"
+          "i16:16:32-i32:32:32-i64:32:32-"
+          "f32:32:32-f64:32:32-v64:32:32-"
+          "v128:32:32-v256:32:32-v512:32:32-v1024:32:32-a0:0:32-n32";
       AddrSpaceMap = &TCEOpenCLAddrSpaceMap;
       UseAddrSpaceMapMangling = true;
     }
