--- llvm-2.9/lib/Target/TargetLibraryInfo.cpp	2011-03-03 16:20:22.000000000 +0200
+++ llvm-2.9-modified/lib/Target/TargetLibraryInfo.cpp	2011-06-06 15:57:36.000000000 +0300
@@ -31,8 +31,8 @@
   if (T.getOS() != Triple::Darwin || T.getDarwinMajorNumber() < 9)
     TLI.setUnavailable(LibFunc::memset_pattern16);
 
-  // iprintf and friends are only available on XCore.
-  if (T.getArch() != Triple::xcore) {
+  // iprintf and friends are only available on XCore and TCE.
+  if (T.getArch() != Triple::xcore && T.getArch() != Triple::tce) {
     TLI.setUnavailable(LibFunc::iprintf);
     TLI.setUnavailable(LibFunc::siprintf);
     TLI.setUnavailable(LibFunc::fiprintf);
